Add isPow2, aligned, and align utility to Bytes.hpp

rwy0717 opened the pull request:
The align utility is for quickly rounding sizes up to a power of two.
Aligning is done everywhere, so it makes sense to have one common
implementation. Note that this is not a general rounding utility, as it
can only round certain specially crafted inputs.

Signed-off-by: Robert Young <rwy0717@gmail.com>

fjeremic commented on include_core/OMR/Bytes.hpp:
```diff
@@ -53,6 +53,25 @@ gibibytes(size_t n)
+inline size_t
+align(size_t size, size_t alignment)
+{
+	return (size + (alignment - 1)) & ~(alignment - 1);
```
Given the "shoot yourself in the foot" likelyhood of someone using this for non-powers of two, can we add an assert in here to check that `alignment` is a power of two rather than relying on the user reading the documentation. We can use something like [1] to check for it.

[1] https://graphics.stanford.edu/~seander/bithacks.html#DetermineIfPowerOf2

0xdaryl replied:
+1.   That query alone should be a utility in this core library somewhere.  The power-of-2 question is already asked in the JIT, so we can eventually replace those questions with whatever becomes available in core.

charliegracie replied:
If this is going to be used by all components, including the GC, which currently has the exact code above I would rather keep the code to a minimum as to not impact the performance in tights loops in the GC that calls this code.  There is tons of complicated code that requires users to read the API documentation before they call it. If preferred we can rename the API to something that implies a power of 2 alignment but I would rather not slow down the code for no reason.

rwy0717 replied:
Yeah, my intention was for this to be a "quick and dirty" call. Especially when you consider the use cases--fast path allocator for example--it makes sense to keep it small. 

I could split this into a safe `align` and an `alignUnsafe` helper. I can also add `isAligned(n, a)` and `isPow2(x)` to the header, while I'm here.

There aren't any assertion utilities in OMR core (yet, AFAIK), so any checks would have to be done with the standard C assert macro. cassert is disabled in release builds, which I think is the behaviour we want, although these kinds of assertions tend to mislead people when debugging.

rwy0717 replied:
Do we want to accept an alignment of zero, or not?

charliegracie replied:
Yes do not accept 0

jdmpapin commented on include_core/OMR/Bytes.hpp:
```diff
@@ -53,6 +53,25 @@ gibibytes(size_t n)
+#endif
+
+/// Round a size up to a multiple of alignment.
+/// Alignment must be a positive multiple of two, EG 1, 2, 4, 8, or 16.
```
Should this say "*power* of two"?

rwy0717 replied:
Ah haha for sure. Will fix.

rwy0717 replied:
Fixed, thanks.

rwy0717 commented:
The planned functions:

* `isPow2(x)` and `isPow2OrZero(x)`
  * determine if x is a power of two.
  * OrZero variant does not check for zero.
* `aligned(n, a)` and `alignedUnsafe(n, a)`
  * determine if `n` is aligned to `a`.
  * The unsafe variant doesn't validate arguments.
* `align(n, a)` and `alignUnsafe(n, a)`
  * round `n` up to `a`.

rwy0717 commented:
The value of `isPow2OrZero` is dubious... unless someone has a specific use case, I won't add it.

rwy0717 commented:
Updated and ready for another look.

I'm concerned that the range checking is too restrictive. Currently it allows alignments or sizes up to `(SIZE_MAX / 2) + 1`. Are we happy with these limits? Or I could just drop the range check all together, if that's what people want.

rwy0717 commented:
I can always do a more flexible overflow check, and nix the max constants.

rwy0717 commented:
changes:
* renamed `alignUnsafe`/`alignedUnsafe` to `alignNoCheck`/`alignedNoCheck`, which is a more conventional naming scheme.
* replaced `size` range check with more permissive overflow check.
* removed` alignment` range check, any `size_t` larger than `ALIGNMENT_MAX` will fail the `isPow2` check anyways.

I left the `UNALIGNED_SIZE_MAX` and `ALIGNMENT_MAX` constants in the header, in case anyone finds them useful.

rwy0717 commented:
Updated the comments. Ready for another look.

charliegracie commented:
@genie-omr build all

youngar commented on fvtest/coretest/TestBytes.cpp:
```diff
@@ -48,4 +48,140 @@ TEST(TestBytes, CompareOneUnitToSmallerUnit)
+	EXPECT_EQ(8, align(7, 8));
+	EXPECT_EQ(8, align(8, 8));
+	EXPECT_EQ(16, align(9, 8));
+	EXPECT_EQ(gibibytes(1), align(gibibytes(1), 4));
```
```suggestion
	EXPECT_EQ(gibibytes(1), align(gibibytes(1), 8));
```

youngar commented:
@genie-omr build all
